File: harness/oracle/cases/chat-activity.ts
/**
 * Oracle case (P4.D140, v4 `735d9408c` — bug 112): the chat-activity
 * chokepoint.
 *
 * Drives the REAL exports of v4's `lib/chat/chat-activity.ts`:
 * `isCharacterAuthoredMessage`, `CHARACTER_AUTHORED_MESSAGE_FILTER`,
 * `chatActivityAt`, `chatActivityTime` and `byChatActivityDesc`.
 *
 * The corpus is v4's own test table (`lib/chat/__tests__/chat-activity.test.ts`
 * — every Staff sender by name, the whisper/silent inclusions, the announcement
 * bubble, the TOOL/SYSTEM roles, the two non-message event types) PLUS the
 * edges v4's table does not state and this port must not guess:
 *   - `systemSender: ''` — the truthiness-vs-`IS NULL` seam between the
 *     in-memory predicate and the SQL mirror (v4 ships both spellings
 *     knowingly); measured here rather than assumed.
 *   - `customAnnouncer: {}` / `null` — `{}` is truthy in JS, so an announcement
 *     bubble with no fields is still an announcement.
 *   - a missing `role` / a lowercase `'user'`.
 *   - `lastMessageAt: ''` — `??` is nullish, so the empty string WINS over
 *     `createdAt` and then reads as time 0.
 *   - `byChatActivityDesc` over a list mixing stamped and never-stamped chats
 *     (the sort v5's Salon list, home, projects and Brahma console all share).
 *
 * The SQL mirror is emitted as v4's own filter object; the Rust side asserts
 * its WHERE-fragment spelling encodes exactly those four clauses (the two
 * spellings are prose in v4 and SQL in v5, so they cannot be byte-diffed).
 *
 * Run from inside the server checkout (pinned worktree — ledger §5.1):
 *   cd /tmp/qt-v4-pin-p4d140-735d9408c
 *   npx tsx ~/source/quilltap-v5/harness/oracle/cases/chat-activity.ts \
 *     > /tmp/oracle-chat-activity.ndjson
 */

import {
  isCharacterAuthoredMessage,
  CHARACTER_AUTHORED_MESSAGE_FILTER,
  chatActivityAt,
  chatActivityTime,
  byChatActivityDesc,
} from '@/lib/chat/chat-activity'
import type { ChatEvent } from '@/lib/schemas/types'

type Row =
  | { kind: 'filter'; id: string; filter: unknown }
  | { kind: 'predicate'; id: string; event: Record<string, unknown>; out: boolean }
  | {
      kind: 'activity'
      id: string
      chat: Record<string, unknown>
      at: string
      time: number
    }
  | { kind: 'sort'; id: string; chats: Record<string, unknown>[]; order: string[] }

const rows: Row[] = []

rows.push({
  kind: 'filter',
  id: 'sql-mirror',
  filter: CHARACTER_AUTHORED_MESSAGE_FILTER,
})

/** A minimal character-authored message; overrides carve out each edge case. */
function msg(overrides: Record<string, unknown> = {}): Record<string, unknown> {
  return {
    type: 'message',
    id: '00000000-0000-4000-8000-000000000001',
    role: 'ASSISTANT',
    content: 'hello',
    createdAt: '2026-01-01T00:00:00.000Z',
    ...overrides,
  }
}

const predicate = (id: string, event: Record<string, unknown>) => {
  rows.push({
    kind: 'predicate',
    id,
    event,
    out: isCharacterAuthoredMessage(event as unknown as ChatEvent),
  })
}

// --- v4's table: counts as activity ----------------------------------------
predicate('assistant-speaking', msg({ role: 'ASSISTANT' }))
predicate('user-speaking', msg({ role: 'USER' }))
predicate('whisper', msg({ targetParticipantIds: ['participant-1'] }))
predicate('silent-message', msg({ isSilentMessage: true }))

// --- v4's table: does not count --------------------------------------------
for (const sender of [
  'lantern',
  'aurora',
  'librarian',
  'concierge',
  'prospero',
  'host',
  'commonplaceBook',
  'ariel',
  'carina',
  'suparna',
  'pascal',
]) {
  predicate(`staff-${sender}`, msg({ systemSender: sender }))
}
predicate(
  'announcement-bubble',
  msg({ role: 'USER', customAnnouncer: { kind: 'custom', displayName: 'The Narrator' } })
)
predicate('tool-role', msg({ role: 'TOOL' }))
predicate('system-role', msg({ role: 'SYSTEM' }))
predicate('context-summary-event', msg({ type: 'context-summary' }))
predicate('system-event', msg({ type: 'system' }))

// --- the edges v4's table does not state -----------------------------------
predicate('system-sender-empty-string', msg({ systemSender: '' }))
predicate('system-sender-null', msg({ systemSender: null }))
predicate('custom-announcer-empty-object', msg({ customAnnouncer: {} }))
predicate('custom-announcer-null', msg({ customAnnouncer: null }))
predicate('empty-whisper-array', msg({ targetParticipantIds: [] }))
predicate('role-lowercase-user', msg({ role: 'user' }))
predicate('role-missing', { type: 'message', id: 'x', content: 'hi', createdAt: '2026-01-01T00:00:00.000Z' })
predicate('type-missing', { id: 'x', role: 'USER', content: 'hi', createdAt: '2026-01-01T00:00:00.000Z' })
predicate('staff-whisper', msg({ systemSender: 'commonplaceBook', targetParticipantIds: ['p1'] }))

// --- chatActivityAt / chatActivityTime -------------------------------------
const activity = (id: string, chat: Record<string, unknown>) => {
  rows.push({
    kind: 'activity',
    id,
    chat,
    at: chatActivityAt(chat as never),
    time: chatActivityTime(chat as never),
  })
}

activity('stamped', {
  lastMessageAt: '2026-05-01T00:00:00.000Z',
  createdAt: '2026-01-01T00:00:00.000Z',
})
activity('null-last-message-at', {
  lastMessageAt: null,
  createdAt: '2026-01-01T00:00:00.000Z',
})
activity('absent-last-message-at', { createdAt: '2026-01-01T00:00:00.000Z' })
activity('unparseable-both', { lastMessageAt: 'not a date', createdAt: 'also not a date' })
// `??` is NULLISH: an empty string wins over createdAt, then reads as time 0.
activity('empty-string-last-message-at', {
  lastMessageAt: '',
  createdAt: '2026-01-01T00:00:00.000Z',
})
// `updatedAt` is never consulted — the whole point of the module.
activity('updated-at-is-ignored', {
  lastMessageAt: null,
  createdAt: '2024-01-01T00:00:00.000Z',
  updatedAt: '2026-06-01T00:00:00.000Z',
})
activity('millisecond-precision', {
  lastMessageAt: '2026-05-01T12:34:56.789Z',
  createdAt: '2026-01-01T00:00:00.000Z',
})

// --- byChatActivityDesc ----------------------------------------------------
const sort = (id: string, chats: Record<string, unknown>[]) => {
  rows.push({
    kind: 'sort',
    id,
    chats,
    order: [...chats].sort(byChatActivityDesc as never).map((c) => c.id as string),
  })
}

sort('loud-before-quiet', [
  { id: 'quiet', lastMessageAt: '2024-01-01T00:00:00.000Z', createdAt: '2023-01-01T00:00:00.000Z' },
  { id: 'loud', lastMessageAt: '2026-01-01T00:00:00.000Z', createdAt: '2023-01-01T00:00:00.000Z' },
])
sort('never-spoken-order-by-creation', [
  { id: 'older', lastMessageAt: null, createdAt: '2024-01-01T00:00:00.000Z' },
  { id: 'newer', lastMessageAt: null, createdAt: '2025-01-01T00:00:00.000Z' },
])
// The mix the Salon list actually sees: a Staff-only chat (now NULL) sinking
// below a stamped one, and `updatedAt` unable to rescue it.
sort('mixed-stamped-and-never', [
  { id: 'staff-only', lastMessageAt: null, createdAt: '2023-01-01T00:00:00.000Z', updatedAt: '2026-08-01T00:00:00.000Z' },
  { id: 'spoken-recently', lastMessageAt: '2026-07-01T00:00:00.000Z', createdAt: '2020-01-01T00:00:00.000Z' },
  { id: 'spoken-long-ago', lastMessageAt: '2024-06-01T00:00:00.000Z', createdAt: '2024-01-01T00:00:00.000Z' },
])
// Ties resolve by sort stability only.
sort('ties-keep-input-order', [
  { id: 'first', lastMessageAt: '2026-01-01T00:00:00.000Z', createdAt: '2020-01-01T00:00:00.000Z' },
  { id: 'second', lastMessageAt: '2026-01-01T00:00:00.000Z', createdAt: '2020-01-01T00:00:00.000Z' },
  { id: 'third', lastMessageAt: '2026-01-01T00:00:00.000Z', createdAt: '2020-01-01T00:00:00.000Z' },
])
// Unparseable timestamps sort as 0 — total order, no NaN.
sort('unparseable-sorts-last', [
  { id: 'broken', lastMessageAt: 'not a date', createdAt: 'nor this' },
  { id: 'fine', lastMessageAt: '2025-01-01T00:00:00.000Z', createdAt: '2020-01-01T00:00:00.000Z' },
])

for (const r of rows) process.stdout.write(JSON.stringify(r) + '\n')

File: crates/quilltap-core/src/chat_activity.rs
//! What counts as *activity* in a chat, and which timestamp the UI shows for it.
//!
//! Port of v4's `lib/chat/chat-activity.ts` (`735d9408c`, bug 112) — the
//! chokepoint that ended three different answers to "what counts as a message"
//! living in one codebase.
//!
//! A chat's `updatedAt` moves whenever anything about the row changes — a
//! generated story background landing, a context summary being folded, a
//! Concierge reroute, a token-cost tally. None of that is the conversation
//! moving forward, so none of it belongs in a "last updated" column the reader
//! scans to find where they left off.
//!
//! `lastMessageAt` is the answer to the question actually being asked: **when
//! did a character last post content?** — the human user or an LLM, speaking as
//! themselves. THE single source of truth for that judgement is
//! [`is_character_authored_message`] here; the SQLite mirror of it is
//! [`CHARACTER_AUTHORED_MESSAGE_FILTER`]. Change one and you must change the
//! other, or the live bump and the backfill will disagree.
//!
//! ## The two spellings are not exactly equivalent — deliberately, as in v4
//!
//! The in-memory predicate tests JS **truthiness** (`if (m.systemSender)`),
//! which lets an empty string through as "absent"; the SQL mirror tests
//! `systemSender IS NULL`, which does not. v4 ships both spellings knowingly —
//! the columns record absence as NULL, so the seam is unreachable in practice —
//! and this port mirrors both rather than unifying them, so the two
//! implementations stay diffable against v4 line for line. The harness pins the
//! `''` edge in the direction v4 actually measures.

use serde_json::Value;

/// Did a character — the human user or an LLM — post this as content?
///
/// Included, deliberately: **whispers** (`targetParticipantIds` non-empty). A
/// character murmuring to one other character is still a character speaking; a
/// room full of whispering shouldn't read as a room gone quiet. (There is no
/// `targetParticipantIds` test here at all — they count by omission.)
///
/// Excluded, deliberately:
/// - **Non-`message` events** (`context-summary`, `system`) — bookkeeping.
/// - **Staff / personified-feature announcements** (`systemSender` set: Lantern,
///   Aurora, Librarian, Concierge, Prospero, Host, Commonplace Book, Ariel,
///   Carina, Suparṇā, Pascal). These persist as `type: 'message'` rows, which is
///   precisely why "any message row" is the wrong test — a background image
///   finishing rendering would otherwise float a months-dead chat to the top of
///   the list.
/// - **Announcement bubbles** (`customAnnouncer` set) — an announcement wearing
///   a name is still an announcement, not the character speaking.
/// - **`SYSTEM` and `TOOL` roles** — a raw tool-result row is machinery, not
///   posted content.
pub fn is_character_authored_message(event: &Value) -> bool {
    if event.get("type").and_then(Value::as_str) != Some("message") {
        return false;
    }
    is_character_authored_parts(
        event.get("role").and_then(Value::as_str).unwrap_or(""),
        event.get("systemSender"),
        event.get("customAnnouncer"),
    )
}

/// The predicate's field-level core, shared by the JSON spelling above and the
/// typed one the write path uses (`ChatEventInput` is already narrowed to the
/// `message` arm there, so the `type` test lives at each caller). Both
/// `system_sender` and `custom_announcer` are tested for **JS truthiness**, as
/// v4's `if (m.systemSender)` / `if (m.customAnnouncer)` are.
pub fn is_character_authored_parts(
    role: &str,
    system_sender: Option<&Value>,
    custom_announcer: Option<&Value>,
) -> bool {
    if role != "USER" && role != "ASSISTANT" {
        return false;
    }
    if js_truthy(system_sender) {
        return false;
    }
    if js_truthy(custom_announcer) {
        return false;
    }
    true
}

/// The SQLite mirror of [`is_character_authored_message`], for indexed lookups
/// that must not load and Zod-validate a whole transcript. v4 spells it as a
/// `QueryFilter` object spread alongside a `chatId`; v5's repositories write SQL
/// directly, so it is the WHERE fragment — one string, so the live lookup and
/// the recompute pass cannot drift.
///
/// `systemSender: null` / `customAnnouncer: null` translate to `IS NULL`, which
/// is how both columns record "absent" (they default to NULL).
pub const CHARACTER_AUTHORED_MESSAGE_FILTER: &str =
    "type = 'message' AND role IN ('USER', 'ASSISTANT') \
     AND systemSender IS NULL AND customAnnouncer IS NULL";

/// The timestamp to sort and display a chat by: when a character last posted,
/// falling back to when the chat was created.
///
/// The fallback is `createdAt`, **not** `updatedAt` — a chat where only the
/// Staff has ever spoken has had no conversational activity at all, and dating
/// it by the last background image regenerated is the very drift this module
/// exists to stop. `createdAt` is the honest, and stable, answer.
///
/// v4's `chat.lastMessageAt ?? chat.createdAt` is **nullish** coalescing: an
/// empty-string `lastMessageAt` wins over `createdAt` (and then reads as time 0
/// through [`chat_activity_time`]). v5's chats arrive as JSON, where an absent
/// key and a JSON `null` are both "nullish". A missing `createdAt`, and a
/// present-but-non-string `lastMessageAt`, are both shapes v4's types forbid
/// (`lastMessageAt` is a TEXT column); they degrade to `""`, i.e. time 0.
pub fn chat_activity_at(chat: &Value) -> &str {
    match chat.get("lastMessageAt") {
        Some(Value::String(s)) => s,
        Some(v) if !v.is_null() => "",
        _ => chat.get("createdAt").and_then(Value::as_str).unwrap_or(""),
    }
}

/// [`chat_activity_at`] as epoch milliseconds, for comparators. An unparseable
/// timestamp reports 0 rather than NaN, so comparators stay total (v4:
/// `Number.isNaN(ms) ? 0 : ms`).
pub fn chat_activity_time(chat: &Value) -> i64 {
    crate::clock::iso_to_ms(chat_activity_at(chat)).unwrap_or(0)
}

/// Newest-activity-first comparator for chat lists (v4 `byChatActivityDesc`).
/// Ties are resolved by sort stability only — use it with a stable sort, as
/// `Array.prototype.sort` is.
pub fn by_chat_activity_desc(a: &Value, b: &Value) -> std::cmp::Ordering {
    chat_activity_time(b).cmp(&chat_activity_time(a))
}

/// JS truthiness over a JSON value — v4's `if (m.systemSender)`.
fn js_truthy(v: Option<&Value>) -> bool {
    match v {
        None | Some(Value::Null) => false,
        Some(Value::Bool(b)) => *b,
        Some(Value::Number(n)) => n.as_f64().map(|f| f != 0.0).unwrap_or(true),
        Some(Value::String(s)) => !s.is_empty(),
        Some(Value::Array(_)) | Some(Value::Object(_)) => true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn an_empty_string_system_sender_is_absent_to_the_predicate_but_not_to_the_sql() {
        // The deliberate seam between the two spellings (module header): the
        // in-memory predicate tests truthiness, the SQL mirror tests IS NULL.
        assert!(is_character_authored_message(&json!({
            "type": "message", "role": "USER", "systemSender": ""
        })));
        assert!(CHARACTER_AUTHORED_MESSAGE_FILTER.contains("systemSender IS NULL"));
    }

    #[test]
    fn an_empty_announcer_object_is_still_an_announcement() {
        // `{}` is truthy in JS — an announcement bubble with no fields is still
        // an announcement.
        assert!(!is_character_authored_message(&json!({
            "type": "message", "role": "USER", "customAnnouncer": {}
        })));
    }

    #[test]
    fn activity_falls_back_to_created_at_never_updated_at() {
        let chat = json!({
            "lastMessageAt": Value::Null,
            "createdAt": "2026-01-01T00:00:00.000Z",
            "updatedAt": "2026-05-01T00:00:00.000Z",
        });
        assert_eq!(chat_activity_at(&chat), "2026-01-01T00:00:00.000Z");
    }
}
